src/compiler/program.ts

```diff
@@ -406,8 +406,12 @@ namespace ts {
             const lineStart = getPositionOfLineAndCharacter(file, i, 0);
             const lineEnd = i < lastLineInFile ? getPositionOfLineAndCharacter(file, i + 1, 0) : file.text.length;
             let lineContent = file.text.slice(lineStart, lineEnd);
-            lineContent = lineContent.replace(/\s+$/g, "");  // trim from end
-            lineContent = lineContent.replace("\t", " ");    // convert tabs to single spaces
+
+            // Trim from end
+            lineContent = lineContent.replace(/\s+$/g, "");
+
+            // Convert all leading tabs to single spaces
+            lineContent = lineContent.replace(/\t/gy, " ");

```

Comment:
Why the `y` flag?

Reply:
I though that only leading tabs in source code (used for indentation) should be replaced by spaces. I was not sure if replacing all tabs would be ok.

```
// example source code with leading tabs and tabs used elsewhere in the line

			private readonly foo: string;			// replace only leading tabs
```
`/\t/gy` prints:
```
   private readonly foo: string;			// replace only leading tabs
```

`/\t/g` prints:
```
   private readonly foo: string;   // replace only leading tabs
```
If you are ok to replace all tabs with spaces I can of course remove the `y` flag.

Reply:
So the iterative nature of `replace` combined with `y=sticky` means that the regex stops matching after the initial run, right? If so, that's pretty subtle.

I'd rather have this explicitly documented, adding "(the sticky flag 'y' means that the regex stops matching after the initial run of tabs)" or something like it.

Also, [the MDN page](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/sticky) says the `g` is ignored for regexes with `y`.

Reply:
To be honest, I'm not so confident with regular expressions that it took me some time to come up with this solution to replace **all leading** tab characters. I was surprised not to find an example where this is done. I've seen the note on MDN as well but without the _global_ flag, only the first tab character was replaced in my test.

```javascript
// input text with two leading tab characters and two tab characters elsewhere in the text
var input="     Some text       containing tabs";

// A:
console.log(input.replace(/\t/g, "LEADING_TAB_"));  
// prints: "LEADING_TAB_LEADING_TAB_Some textLEADING_TAB_LEADING_TAB_containing tabs"

// B:
console.log(input.replace(/\t/, "LEADING_TAB_"));  
// prints: "LEADING_TAB_   Some text       containing tabs"

// C:
console.log(input.replace(/^(\t)*/, "LEADING_TAB_"));  
// prints: "LEADING_TAB_Some text      containing tabs"

// D:
console.log(input.replace(/\t/gy, "LEADING_TAB_"));  
// prints: "LEADING_TAB_LEADING_TAB_Some text      containing tabs"

// E:
console.log(input.replace(/\t/y, "LEADING_TAB_"));  
// prints: "LEADING_TAB_   Some text       containing tabs"
```
I've created a [jsfiddle](https://jsfiddle.net/onebcvu4/2/) and asked about **why this works** on [stackoverflow](https://stackoverflow.com/questions/64982135/regex-to-replace-all-leading-tab-characters-each-one-with-a-single-space). 

While writing this comment, my question was [answered](https://stackoverflow.com/a/64982184/681803).

Now I just need to find a good comment that clarifies the use of both _sticky_ and _global_ for this line.

Does this do it for you?

```javascript
// Convert all leading tabs to single spaces
// using both g and y flag to repeatedly match a \t at lastIndex (default: 0) and replace it until it runs out of \t at lastIndex
lineContent = lineContent.replace(/\t/gy, " "); 
```

Reply:
Not sure though about what's the max number of characters per line at Typescript and guideline for multi line comments.